src/merkle.rs

```diff
@@ -232,27 +276,144 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>> MerkleTree<T, A, K> {
                         .write()
                         .unwrap()
                         .copy_from_slice(&hashed_nodes_as_bytes, write_start + write_delta)?;
+
+                    Ok(())
+                })?,
+        )
+    }
+
+    fn process_disk_layer(
+        data: &mut K,
+        file: &std::fs::File,
+        width: usize,
+        level: usize,
+        read_start: usize,
+        write_start: usize,
+    ) -> Result<()> {
+        let data_lock = Arc::new(RwLock::new(data));
+
+        Ok(
+            Vec::from_iter((read_start..read_start + width).step_by(BUILD_CHUNK_NODES))
+                .par_iter()
+                .try_for_each(move |&chunk_index| -> Result<()> {
+                    let chunk_size =
+                        std::cmp::min(BUILD_CHUNK_NODES, read_start + width - chunk_index);
+
+                    let chunk_nodes = {
+                        // Read everything taking the lock once.
+                        data_lock
+                            .read()
+                            .unwrap()
+                            .read_range(chunk_index..chunk_index + chunk_size)?
+                    };
+
+                    // We write the hashed nodes to the next level in the position that
+                    // would be "in the middle" of the previous pair (dividing by 2).
+                    let write_delta = (chunk_index - read_start) / 2;
+
+                    let nodes_size = (chunk_nodes.len() / 2) * T::byte_len();
+                    let hashed_nodes_as_bytes = chunk_nodes.chunks(2).fold(
+                        Vec::with_capacity(nodes_size),
+                        |mut acc, node_pair| {
+                            let h = A::default().node(
+                                node_pair[0].clone(),
+                                node_pair[1].clone(),
+                                level,
+                            );
+                            acc.extend_from_slice(h.as_ref());
+                            acc
+                        },
+                    );
+
+                    // Check that we correctly pre-allocated the space.
+                    let chunk_size =
+                        std::cmp::min(BUILD_CHUNK_NODES, read_start + width - chunk_index);
+
+                    let hashed_nodes_as_bytes_len = hashed_nodes_as_bytes.len();
+                    debug_assert_eq!(hashed_nodes_as_bytes_len, chunk_size / 2 * T::byte_len());
+
+                    let map_write_start =
+                        (write_start * T::byte_len() + write_delta * T::byte_len()) as u64;
+
+                    // Safety: this operation is safe becase it's a limited
+                    // writable region that is not overlapped with any
+                    // other mapping that could happen in parallel.
+                    let mut mmap = unsafe {
```

Comment:
wouldn't it be more efficient to create a single memmap at the beginning, and then call `.par_chunks_mut` on it, to write to it?

Reply:
I wasn't able to get this working just yet, but will play with.

Reply:
I pushed a commit that switches it over. The main trick is to know that hat `mmap` derefs into `&mut [u8]`, which can be iterated over using `par_chunks_mut`

Reply:
Nice, thanks!